flutter/lib/src/screenshot/recorder.dart

```diff
@@ -47,9 +58,31 @@ class ScreenshotRecorder {
       // On Android, the desired resolution (coming from the configuration)
       // is rounded to next multitude of 16 . Therefore, we scale the image.
       // On iOS, the screenshot resolution is not adjusted.
-      final srcWidth = renderObject.size.width;
-      final srcHeight = renderObject.size.height;
-      final pixelRatio = config.getPixelRatio(srcWidth, srcHeight);
+      final srcWidth = renderObject.size.width.toInt();
+      final srcHeight = renderObject.size.height.toInt();
+
+      final int targetWidth;
+      final int targetHeight;
+
+      final double pixelRatio;
+
+      // If width or height is not set, we use the device resolution.
+      if (config.width == null || config.height == null) {
+        final binding = options.bindingUtils.instance!;
+
+        // ignore: deprecated_member_use
+        targetHeight = binding.window.physicalSize.height.toInt();
+        // ignore: deprecated_member_use
+        targetWidth = binding.window.physicalSize.width.toInt();
+
+        pixelRatio = max(targetWidth, targetHeight) / max(srcWidth, srcHeight);
+      } else {
+        pixelRatio =
+            max(config.width!, config.height!) / max(srcWidth, srcHeight);
+
+        targetWidth = (srcWidth * pixelRatio).toInt();
+        targetHeight = (srcHeight * pixelRatio).toInt();
+      }
```

Comment:
In #2409 I noticed that you removed the line `options.screenshotQuality = SentryScreenshotQuality.low;` in the `main.dart`. Was this intentional or by accident?

For the the Quality `low`, `medium`, and `high` everything works with your suggested changes in #2409, but for `full` the output image is about the size as the `low` setting, as it uses the `logical resolution` and not the `device resolution` therefore it would break the previous behaviour.

To capture screenshots with quality `full` and keep the previous functionality, we need to get either the `device resolution` or the `devicePixelRatio`. As both values can change, for `Web` and `Desktop` applications, due to resizing of the window, setting the `ScreenshotRecorderConfig` to a fixed value at app start does not work. As it will stay the same even after resizing the window. Therefore the calculation needs to be done, before every screenshot capture. But this approach is only working if the `widget` is covering the complete device display, which is no longer a must.

For me it looks like, the best/only/easiest solution would be to deprecate the `full` setting in the `SentryScreenshotQuality` due to the change of the approach of generating the screenshot. This would also make sense, regarding the possibility, of only capturing a screenshot of a specific widget, which is neither related to the `device resolution` nor to the `device orientation`.
By ignoring the `full` setting, the user will still get screenshots, but with a lower resolution, as with the previous setting, and therefore the change is not that big.

@vaind @buenaflor WDYT? Should we deprecate the `full` or do you have another suggestion?

Should we exchange the `full`, with `extraHigh`? But this would be a breaking change, because it will no longer return `null` and the `enum name` changes and the resolution will also change.

Reply:
> In #2409 [here](https://github.com/getsentry/sentry-dart/pull/2409#discussion_r1842042755) I noticed that you removed the line `options.screenshotQuality = SentryScreenshotQuality.low;` in the `main.dart`. Was this intentional or by accident?

intentional, I was thinking we should be testing the default value for quality when testing manually, so that it's the same as what users see by default.
 
> For the the Quality `low`, `medium`, and `high` everything works with your suggested changes in #2409, but for `full` the output image is about the size as the `low` setting, as it uses the `logical resolution` and not the `device resolution` therefore it would break the previous behaviour.

Interesting, it worked for me as expected on Windows. Which platforms have you tried this on so far?


> To capture screenshots with quality `full` and keep the previous functionality, we need to get either the `device resolution` or the `devicePixelRatio`. 

That is a very good point. We should grab devicePixelRatio and take it into account when calculating the pixelRatio based on constraints: https://github.com/getsentry/sentry-dart/pull/2409/files#diff-b603ed629ae7bb928af4f892324700593bfeb22ecff70a571e58345034b881a4R61
That should be a small change, I can have a look if you want. 

> As both values can change, for `Web` and `Desktop` applications, due to resizing of the window, setting the `ScreenshotRecorderConfig` to a fixed value at app start does not work. As it will stay the same even after resizing the window. Therefore the calculation needs to be done, before every screenshot capture. But this approach is only working if the `widget` is covering the complete device display, which is no longer a must.

For screenshots, we don't read or set window size at the start, we read it when capturing the screnshot: https://github.com/getsentry/sentry-dart/pull/2409/files#diff-b603ed629ae7bb928af4f892324700593bfeb22ecff70a571e58345034b881a4R59-R60. For replay, this shall change too when we support orientation change.

The configuration for `full` has both `width` and `height` as null, therefore, the calculated pixelRatio (as in the original code or in the revert PR, not in the current version here) would be 1.0 so that's not a problem. 

> For me it looks like, the best/only/easiest solution would be to deprecate the `full` setting in the `SentryScreenshotQuality` due to the change of the approach of generating the screenshot. This would also make sense, regarding the possibility, of only capturing a screenshot of a specific widget, which is neither related to the `device resolution` nor to the `device orientation`. By ignoring the `full` setting, the user will still get screenshots, but with a lower resolution, as with the previous setting, and therefore the change is not that big.
> 
> @vaind @buenaflor WDYT? Should we deprecate the `full` or do you have another suggestion?
> 
> Should we exchange the `full`, with `extraHigh`? But this would be a breaking change, because it will no longer return `null` and the `enum name` changes and the resolution will also change.

I don't think of the screenshot quality as the function of window size, but of the size of the captured screenshot, which is always coming from the ScreenshotWidget size. I'd keep the setting, with my above suggestion to change the pixelRatio calculation, it should work the same as before.

Reply:
> > In #2409 [here](https://github.com/getsentry/sentry-dart/pull/2409#discussion_r1842042755) I noticed that you removed the line `options.screenshotQuality = SentryScreenshotQuality.low;` in the `main.dart`. Was this intentional or by accident?
> 
> intentional, I was thinking we should be testing the default value for quality when testing manually, so that it's the same as what users see by default.

But we want to keep it in the example app right? You removed it only for testing purposes right?

> > For the the Quality `low`, `medium`, and `high` everything works with your suggested changes in #2409, but for `full` the output image is about the size as the `low` setting, as it uses the `logical resolution` and not the `device resolution` therefore it would break the previous behaviour.
> 
> Interesting, it worked for me as expected on Windows. Which platforms have you tried this on so far?

My last test was with the `Pixel 7 Pro API 35` emulator on my mac.

> > To capture screenshots with quality `full` and keep the previous functionality, we need to get either the `device resolution` or the `devicePixelRatio`.
> 
> That is a very good point. We should grab devicePixelRatio and take it into account when calculating the pixelRatio based on constraints: https://github.com/getsentry/sentry-dart/pull/2409/files#diff-b603ed629ae7bb928af4f892324700593bfeb22ecff70a571e58345034b881a4R61 That should be a small change, I can have a look if you want.

But then we don't need the `getPixelRatio` anymore, because we will get the `devicePixelRatio` from [`options.bindingUtils.instance.window.devicePixelRatio`](https://api.flutter.dev/flutter/dart-ui/FlutterView/devicePixelRatio.html) and this is fixed for every device and not related to the orientation and the size of the `window/widget`. So we can set this, on app start.
In addition the whole `ScreenshotRecorderConfig` will not be needed anymore, because the `width` and `height` properties were only used, for calculating the `ratio` but this approach is not needed anymore.

> > As both values can change, for `Web` and `Desktop` applications, due to resizing of the window, setting the `ScreenshotRecorderConfig` to a fixed value at app start does not work. As it will stay the same even after resizing the window. Therefore the calculation needs to be done, before every screenshot capture. But this approach is only working if the `widget` is covering the complete device display, which is no longer a must.
> 
> For screenshots, we don't read or set window size at the start, we read it when capturing the screnshot: https://github.com/getsentry/sentry-dart/pull/2409/files#diff-b603ed629ae7bb928af4f892324700593bfeb22ecff70a571e58345034b881a4R59-R60. For replay, this shall change too when we support orientation change.
> 
> The configuration for `full` has both `width` and `height` as null, therefore, the calculated pixelRatio (as in the original code or in the revert PR, not in the current version here) would be 1.0 so that's not a problem.

Depends on the definition of a `problem`.
I just tested it with the current `main` branch, and choose the `full` setting and it resulted in an image with the size of `1440 × 3036 pixels` which matches the screen resolution of the `pixel 7 pro` emulator(`1440x3120 pixels`).

As already written, with the `ratio`of 1.0 the resulting output image has the size of the `logical resolution`, which is about the size as the `low` setting. Therefore it will break the current behaviour, which the user gets in the current release. 
> 
> > For me it looks like, the best/only/easiest solution would be to deprecate the `full` setting in the `SentryScreenshotQuality` due to the change of the approach of generating the screenshot. This would also make sense, regarding the possibility, of only capturing a screenshot of a specific widget, which is neither related to the `device resolution` nor to the `device orientation`. By ignoring the `full` setting, the user will still get screenshots, but with a lower resolution, as with the previous setting, and therefore the change is not that big.
> > @vaind @buenaflor WDYT? Should we deprecate the `full` or do you have another suggestion?
> > Should we exchange the `full`, with `extraHigh`? But this would be a breaking change, because it will no longer return `null` and the `enum name` changes and the resolution will also change.
> 
> I don't think of the screenshot quality as the function of window size, but of the size of the captured screenshot, which is always coming from the ScreenshotWidget size. I'd keep the setting, with my above suggestion to change the pixelRatio calculation, it should work the same as before.

At least the behaviour of the current stable, results in a screenshot with the device resolution. But I agree, that this behaviour is not 100% compatible with our new approach.

Reply:
> But then we don't need the `getPixelRatio` anymore, because we will get the `devicePixelRatio` from [`options.bindingUtils.instance.window.devicePixelRatio`](https://api.flutter.dev/flutter/dart-ui/FlutterView/devicePixelRatio.html) and this is fixed for every device and not related to the orientation and the size of the `window/widget`. So we can set this, on app start.
> In addition the whole `ScreenshotRecorderConfig` will not be needed anymore, because the `width` and `height` properties were only used, for calculating the `ratio` but this approach is not needed anymore.

We need to take the configuration into account alongside the devicePixelRatio, otherwise anything but full-resolution images would be captured incorrectly.

Reply:
> But we want to keep it in the example app right? You removed it only for testing purposes right?

Not really, I was thinking of removing that setting in the sample app. Besides being an example, it's also an app used for manual testing. My thinking is that we should be testing the default here.

Reply:
> Depends on the definition of a `problem`. I just tested it with the current `main` branch, and choose the `full` setting and it resulted in an image with the size of `1440 × 3036 pixels` which matches the screen resolution of the `pixel 7 pro` emulator(`1440x3120 pixels`).
> 
> As already written, with the `ratio`of 1.0 the resulting output image has the size of the `logical resolution`, which is about the size as the `low` setting. Therefore it will break the current behaviour, which the user gets in the current release.

You're right. I've written this part of the response before going back to the devicePixelRatio - that needs to be considered.